File: plot/plot_mcp/slug.py
"""Folder-path slug for node → folder mapping (v0.7).

Produces the relative path used by ``SketchNode.folder_path``. The rule is
deliberately mechanical so both sides (Python server, TypeScript client)
can compute it independently: ``workspace/{canvas}/{kind}-{label-slug}/``.

Korean and other non-ASCII characters are preserved — they work fine on
every modern filesystem and they read better than transliteration.
"""

from __future__ import annotations

import re

# Non-Korean letter ranges we want to preserve. Keep it inclusive: Latin
# letters+digits, Hangul (AC00–D7AF for syllables, 1100–11FF for jamo,
# A960–A97F extended), CJK in case user mixes. Everything else collapses
# to a single dash.
_PRESERVE_RE = re.compile(
    r"[^a-z0-9"
    r"가-힯"
    r"ᄀ-ᇿ"
    r"ꥠ-꥿"
    r"぀-ヿ"  # Hiragana/Katakana — may as well
    r"一-鿿"  # CJK Unified
    r"]+",
    re.IGNORECASE,
)


def slugify(text: str) -> str:
    """Lowercase ASCII letters/digits, preserve Korean/CJK, collapse
    everything else to single dashes, trim leading/trailing dashes."""
    lowered = text.lower().strip()
    replaced = _PRESERVE_RE.sub("-", lowered)
    return replaced.strip("-")


def folder_slug(kind: str, label: str, canvas: str = "core") -> str:
    """Compose a ``workspace/{canvas}/{kind}-{label}`` relative folder path.

    Called by both server and client to agree on the default location for a
    new node's folder. The server then uniquifies if taken.
    """
    kind_slug = slugify(kind) or "node"
    label_slug = slugify(label)
    stem = f"{kind_slug}-{label_slug}" if label_slug else kind_slug
    canvas_slug = slugify(canvas) or "core"
    return f"workspace/{canvas_slug}/{stem}"

File: plot/viewer/src/lib/slug.ts
/**
 * Folder-path slug for node → folder mapping (v0.7).
 *
 * Mirrors ``plot_mcp/slug.py`` so the viewer can compute the default
 * folder client-side without an extra round-trip. Server will still
 * uniquify on POST /api/folders if the name is already taken.
 */

// Preserve Latin letters/digits, Hangul + CJK, collapse the rest to dashes.
const PRESERVE_RE =
  /[^a-z0-9가-힯ᄀ-ᇿꥠ-꥿぀-ヿ一-鿿]+/gi;

export function slugify(text: string): string {
  const lowered = text.toLowerCase().trim();
  const replaced = lowered.replace(PRESERVE_RE, "-");
  return replaced.replace(/^-+|-+$/g, "");
}

export function folderSlug(kind: string, label: string, canvas = "foundation"): string {
  const kindSlug = slugify(kind) || "node";
  const labelSlug = slugify(label);
  const stem = labelSlug ? `${kindSlug}-${labelSlug}` : kindSlug;
  const canvasSlug = slugify(canvas) || "foundation";
  return `${canvasSlug}/${stem}`;
}
